### ISSUE TITLE
About code for pretraining

### ISSUE THREAD

**Issue by volgachen** [encounter1997/FP-DETR#4]
Excuse me, do you have any plan to release codes or instructions for pretraining?

**Comment by encounter1997** [comment]
Sorry that we do not have the plan to release the code for pre-training, but it can be easily implemented by replacing the [model construction function](https://github.com/facebookresearch/deit/blob/e41bf4b4332c3c1191750a6a56ea1f86f36a37bc/main.py#L254) in the [DeiT code](https://github.com/facebookresearch/deit) with [our model construction function](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/tools/train.py#L158).

Hope this can help you and feel free to ask anything if you have difficulties in implementing the pre-training code.

**Comment by volgachen** [comment]
Thank you for your response.
I guess I should modify `query_shape` into `(1,1)`. Is there any other configs I should notice?

**Comment by encounter1997** [comment]
Taking [fp-detr-base-in1k.py](https://github.com/encounter1997/FP-DETR/blob/main/configs/fp_detr/fp-detr-base_in1k.py) for example, there are several parts that should be modified in the config, as follows:
1. Only the [model definition](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L4) is needed;
2. The [self-attn](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L55) and the corresponding norm in encoder2 should be removed, and the [operation order](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L70) should be updated.
3. [return_intermediate](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L50) should be updated since deep supervision is not used during pre-training. The code in the DeiT project may also need to be changed slightly, to obtain the class token from the output sequence for loss computation.
4. [num_classes](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L25) should be 1000 for ImageNet classification.

**Comment by volgachen** [comment]
> 2. The [self-attn](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L55) and the corresponding norm in encoder2 should be removed, and the [operation order](https://github.com/encounter1997/FP-DETR/blob/05aa706f9e65d9b835b28fb3b19e3c39726d29db/configs/fp_detr/fp-detr-base_in1k.py#L70) should be updated.

I suppose the `self-attn` you mentioned in point 2 is actually [prompt_self_attn](https://github.com/encounter1997/FP-DETR/blob/main/mmcv_custom/cnn/bricks/transformer.py#L480)?

**Comment by encounter1997** [comment]
Yes, that's right.

**Comment by volgachen** [comment]
Thank you! It seems to be right now.

**Comment by volgachen** [comment]
I find that there is a learning rate decay for `sampling_offsets` in the training configuration for detection.
How do you handle with `sampling_offsets` in the pretraining process?

**Comment by encounter1997** [comment]
We did not carefully tune the learning rate for `sampling_offsets `and `reference_points `during pre-training, and simply set their learning rate the same as other parameters in the transformer encoder. Tuning the learning rate may lead to better pre-training results, but we didn't try.
